### PR TITLE
Fall back to UTF-8 for local UTF-8 text files in the browser

### LINKED ISSUE #12432: Built-in browser renders UTF-8 local text files as mojibake (file:///…/notes.md)
## Bug description

Opening a local UTF-8 text file in a browser pane renders mojibake. A `notes.md` file containing `# 산책의 즐거움` shows up as `# �곗콉�� 利먭굅��`.

Chrome renders the very same file correctly:

```
$ file -I /tmp/notes.md
/tmp/notes.md: text/plain; charset=utf-8

$ "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome" --headless --dump-dom file:///tmp/notes.md
<html>…<pre …># 산책의 즐거움
```

## Steps to reproduce

1. `printf '# 산책의 즐거움\n' > /tmp/notes.md`
2. Open `file:///tmp/notes.md` in a cmux browser pane
3. The text renders as mojibake

## Root cause

A `file://` navigation carries no `Content-Type` charset, and plain text has no `<meta charset>`, so WebKit decodes the document with its locale-derived fallback encoding (`WKPreferences._defaultTextEncodingName`).

Confirmed with a standalone WKWebView harness on this machine (Korean system locale):

| how the file is loaded | `document.characterSet` | `document.contentType` | rendered |
| --- | --- | --- | --- |
| `loadFileURL(notes.md)` — what cmux does | `EUC-KR` | `text/markdown` | `# �곗콉�� 利먭굅��` |
| same bytes with `characterEncodingName: "utf-8"` | `UTF-8` | `text/plain` | `# 산책의 즐거움` |

A pristine `WKPreferences` reports `ks_c_5601-1987` as the fallback in this locale. A US-English system gets `windows-1252` instead, so this is not Korean-specific: any non-ASCII UTF-8 local text file mojibakes in any locale whose fallback is not UTF-8.

The path is `browserLoadRequest` → `webView.loadFileURL(...)` in `Sources/Panels/BrowserPanel.swift`, so it covers every local text file the browser displays inline (`.md`, `.txt`, `.csv`, `.py`, …). Local HTML that declares `<meta charset>` is unaffected, because a declared charset always wins over the fallback.

## Expected behavior

A UTF-8 local text file renders as UTF-8.

## Notes on fixing

Two obvious approaches do not hold up — both measured in the same harness:

- `webView.load(data:mimeType:characterEncodingName:baseURL:)` decodes correctly, but it replaces the current back/forward entry instead of pushing one (`canGoBack` stays `false`), so navigating to a local file loses the previous page. A reload then re-fetches the file through the normal path and the mojibake returns.
- `loadSimulatedRequest(_:response:responseData:)` does keep history, but a reload re-fetches the file the same way, so the page mojibakes again after ⌘R.

What does work is choosing WebKit's fallback encoding per main-frame navigation: switch `_defaultTextEncodingName` to UTF-8 only when the navigation target is a local file whose own bytes decode as UTF-8, and leave WebKit's locale fallback in place otherwise. Verified: remote pages are untouched, a genuinely EUC-KR local file still renders, a declared `<meta charset>` still wins, and reload / back / forward all keep the right encoding.

A PR with that fix follows.

## Environment

- cmux version: 0.64.22 (102)
- macOS: 15.3 (24D60), Apple Silicon
- System locale: ko_KR.UTF-8

### PR DESCRIPTION
**Pull request by zziri**
Fixes #12432

## Summary

**What changed?** A main-frame navigation to a local file now picks WebKit's fallback text encoding: UTF-8 when the target file's own leading bytes decode as UTF-8, and WebKit's locale fallback otherwise. New `BrowserLocalFileTextEncoding` owns that decision (sniff + `_defaultTextEncodingName`), and `BrowserNavigationDelegate` calls it just before allowing the navigation.

**Why?** WebKit derives a document's fallback encoding from the system locale when the document declares no charset, and a local text file cannot declare one — a `file://` navigation carries no `Content-Type` header, and plain text has no `<meta charset>`. So a UTF-8 `notes.md` opened in a browser pane decodes through `ks_c_5601-1987` on a Korean Mac (or `windows-1252` on a Western one) and renders as mojibake: `# 산책의 즐거움` shows up as `# �곗콉�� 利먭굅��`. Chrome renders the same file correctly. Details and measurements are in #12432.

Sniffing the file, rather than forcing UTF-8 for all local files, is what keeps this from making anything worse: a local file that really is EUC-KR keeps decoding through the locale fallback. Remote pages are never touched, so a charset-less legacy page renders exactly as it does today, and a declared `<meta charset>` still wins everywhere. Because the fallback is re-chosen on every main-frame navigation, reload and back/forward keep the encoding their page was first rendered with.

Two loads that carry an explicit charset were measured and rejected:

- `load(data:mimeType:characterEncodingName:baseURL:)` decodes correctly but replaces the current back/forward entry instead of pushing one (`canGoBack` stays `false`), so navigating to a local file loses the previous page.
- `loadSimulatedRequest(_:response:responseData:)` keeps history, but — like the above — a reload re-fetches the file through the normal path and the mojibake comes back on ⌘R.

Two commits per `AGENTS.md`: the first adds the failing coverage, the second the fix.

## Testing

I could not build the app or run `cmuxTests` in my environment (Xcode 26 / GhosttyKit toolchain not available here), so **the added tests have not been executed** — CI is the first run. What I did verify: I compiled `Sources/Panels/BrowserLocalFileTextEncoding.swift` as-is into a standalone WKWebView harness that drives the same `decidePolicyFor navigationAction` hook this PR adds, on macOS 15.3 with a `ko_KR` system locale, and measured `document.characterSet` plus the rendered text for each case.

| case | before | after |
| --- | --- | --- |
| UTF-8 `notes.md` (Korean) | `EUC-KR`, mojibake | `UTF-8`, correct |
| …after ⌘R and after back/forward | `EUC-KR`, mojibake | `UTF-8`, correct |
| local file that really is EUC-KR | `EUC-KR`, correct | `EUC-KR`, correct (unchanged) |
| local HTML declaring `<meta charset="euc-kr">` | `EUC-KR`, correct | `EUC-KR`, correct (unchanged) |
| remote page | locale fallback | locale fallback (unchanged) |

The sniff was also exercised directly: UTF-8 file → `true`; EUC-KR bytes → `false`; PNG header → `false`; a UTF-8 file whose 64 KiB sniff window splits a multi-byte scalar → `true`; empty file, missing file, `https://` URL, `nil` → `false`.

Added tests mirror those cases: three behavior tests that open local files in a `BrowserPanel` and assert `document.characterSet` and the rendered text, plus four unit tests for the sniff.

## Demo Video

- Video URL or attachment: not included — this is a text-rendering fix; the before/after table above is the observable behavior, and the added tests assert it.

## Checklist

- [x] I tested the change locally (standalone WKWebView harness linking the new file; the app test suite could not be built in my environment — see Testing)
- [x] I added or updated tests for behavior changes
- [ ] I updated docs/changelog if needed (no user-facing strings or docs affected)

---
<a href="https://app.blacksmith.sh/manaflow-ai/codesmith/cmux/pr/12433"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1791818461&installation_model_id=21920&pr_number=12433&repository=manaflow-ai%2Fcmux&return_to=https%3A%2F%2Fgithub.com%2Fmanaflow-ai%2Fcmux%2Fpull%2F12433&signature=caec89a3d8c0dd24961a95f2fde95fb8cfdef7ba8a3d2de8fce7aefc78027bd0"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

---
## Summary by cubic
Fixes #12432: local UTF-8 text files opened in the browser pane now decode as UTF-8 instead of WebKit's locale fallback, which turned UTF-8 content into mojibake on non-UTF-8 system locales.

**Bug Fixes**
- New `BrowserLocalFileTextEncoding` sniffs the file's leading 64 KiB and picks UTF-8 only when the bytes decode as UTF-8.
- When the sniff window splits a scalar, the file is retried without that scalar; this is allowed only when the file continues past the window and the tail is a genuinely cut-short scalar, so malformed prefixes (overlong encodings, surrogates, scalars past U+10FFFF) and invalid tails still fail.
- `BrowserNavigationDelegate` applies the fallback before every main-frame navigation, so reload and back/forward keep the page's original rendering.
- Remote pages and documents with a declared `<meta charset>` are untouched, even when the file's bytes decode as UTF-8.
- Tests cover UTF-8 Markdown and plain text, an EUC-KR file, and sniff edge cases like truncated and malformed scalars, invalid tails, empty/missing files, and remote URLs.

<sup>Written for commit 06a77877285ca97b3b759c32a94661272a0d7afd. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/manaflow-ai/cmux/pull/12433?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved text encoding detection for local files, helping UTF-8 Markdown and text files display correctly.
  * Added support for UTF-8 files with truncated trailing characters.
  * Preserved explicitly declared document encodings, including non-UTF-8 HTML character sets.
  * Maintained existing fallback behavior for remote, empty, missing, unreadable, and non-UTF-8 files.
  * Improved handling of incomplete or invalid multi-byte characters at file boundaries.

### CODE REVIEW

File: Sources/Panels/BrowserLocalFileTextEncoding.swift [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        defer { try? handle.close() }
+        guard var prefix = try? handle.read(upToCount: sniffedByteCount),
+              !prefix.isEmpty else {
+            return false
+        }
+
+        for _ in 0...maximumTruncatedScalarBytes {
+            if String(data: prefix, encoding: .utf8) != nil { return true }
+            guard prefix.count > 1 else { return false }
+            prefix = prefix.dropLast()
```
[the remark this answers is not in the document]
Comment (zziri) [reply Sources/Panels/BrowserLocalFileTextEncoding.swift]: Confirmed and fixed in 3d0e794. A trailing scalar is now dropped only when the prefix filled the sniff window (so the file really does continue past it) and only when the tail is a lead byte whose scalar is genuinely cut short; a byte no scalar can start with now fails the sniff outright. Measured against the real file: `[0x41, 0xFF]` and a window-filling file ending in `0xFF` both return `false`, while a window that splits a 3- or 4-byte scalar still returns `true`. Regression tests added for all three.

File: cmuxTests/BrowserConfigTests.swift [hunk 2]
```diff
@@ -6023,3 +6023,168 @@ final class BrowserOmnibarFocusPolicyTests: XCTestCase {
...
+
+    /// An HTML page that declares its own charset must keep it: the fallback
+    /// only fills in for documents that declare nothing.
+    func testDeclaredCharsetStillWinsForLocalHTML() throws {
+        let directory = try makeTemporaryDirectory()
+        let file = directory.appendingPathComponent("legacy.html")
+        // "한글" in EUC-KR, which is not valid UTF-8.
+        var bytes = Data("<html><head><meta charset=\"euc-kr\"></head><body>".utf8)
+        bytes.append(contentsOf: [0xC7, 0xD1, 0xB1, 0xDB])
+        bytes.append(contentsOf: Data("</body></html>".utf8))
```
[the remark this answers is not in the document]
Comment (zziri) [reply cmuxTests/BrowserConfigTests.swift:6264]: Good catch — that test only proved a non-UTF-8 page keeps rendering, so it is now named for that, and 3d0e794 adds the missing case alongside it: `declared.html` carries `<meta charset="windows-1252">` over the bytes `0xC2 0xA3`, which are valid UTF-8, so the sniff does select the UTF-8 fallback (asserted) and WebKit must still decode through the declared charset. Verified against a real WKWebView: `document.characterSet` is `windows-1252` and the body renders `Â£`.

**Comment by zziri** [comment]
I have read the CLA Document v2.2 and I hereby sign the CLA

File: Sources/Panels/BrowserLocalFileTextEncoding.swift [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                continuationByteCount += 1
+                guard continuationByteCount <= maximumTruncatedScalarContinuationBytes else {
+                    return nil
+                }
+                continue
+            }
+
+            guard let scalarByteCount = scalarByteCount(forLeadByte: byte) else { return nil }
+            let presentByteCount = continuationByteCount + 1
+            return presentByteCount < scalarByteCount ? presentByteCount : nil
```
[the remark this answers is not in the document]
Comment (zziri) [reply Sources/Panels/BrowserLocalFileTextEncoding.swift]: Right — those are malformed rather than cut short, and the check only read the lead byte. Fixed in b7d57a0: when the second byte is present it now has to fall in the range its lead byte allows (`0xE0`→`A0...BF`, `0xED`→`80...9F`, `0xF0`→`90...BF`, `0xF4`→`80...8F`), while a lone lead byte stays eligible since the bytes past the window can still complete it. Measured against real files at the window boundary: `E0 80`, `ED A0`, `F0 80`, `F4 90` and `C0 80` now return `false`; `E0`, `ED`, `F4` alone and `E0 A0`, `ED 80`, `F0 90`, `F4 8F` still return `true`. Both sets are now tests.

**Comment by zziri** [comment]
Rebased onto main (ff4428f3) — the conflict was in `cmux.xcodeproj/project.pbxproj`, where `BrowserLocalFileFinderMenu.swift` landed at the same sorted position as this branch's new file. Both entries are kept.

One thing worth flagging: the last commit is a pure `scripts/normalize-pbxproj.py` run, and the three lines it moves are not mine. Three `CloudWorkspaceRenameService+Directory.swift` entries on main sit outside the order that script enforces, so `scripts/check-pbxproj.sh` — which `ci.yml` runs — fails on main as it stands, and on anything branched from it. I kept it in its own commit so it is easy to see that nothing is added, removed, or renamed, and easy to drop if you would rather fix it separately on main.

**Comment by zziri** [comment]
Rebased onto main (01cfdecc) — same `project.pbxproj` collision as before, both entries kept.

The normalization commit is gone: main is normalized again, so it had nothing left to do and I dropped it. The branch is back to the four commits that belong to this change, and `scripts/check-pbxproj.sh` passes on it as it stands.
